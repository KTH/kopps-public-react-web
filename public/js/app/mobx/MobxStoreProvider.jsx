/* eslint no-use-before-define: ["error", "nofunc"] */

import React from 'react'

import { createMobxManager } from './manager'

export default MobxStoreProvider

/**
 * React-component which prepares a MobX-store
 * so that its data can be used with the useStore() hook
 * of any other component embedded into this Provider.
 *
 * @param {object} props
 * @param {function} props.initCallback
 * @param {string} [props.storeId]
 * @param {object} props.children
 *
 */
function MobxStoreProvider(props) {
  const { initCallback, storeId, children } = props

  const manager = createMobxManager(storeId || 'default', initCallback)

  const { Provider: ReactContextProvider, value: mobxStore } = manager.getProviderComponentAndValue()

  return <ReactContextProvider value={mobxStore}>{children}</ReactContextProvider>
}
