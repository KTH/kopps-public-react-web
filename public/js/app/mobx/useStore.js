/* eslint no-use-before-define: ["error", "nofunc"] */

import createApplicationStore from '../stores/createApplicationStore'
import { getMobxManager } from './manager'

export default useStore

/**
 * Hook which can be used by any React component
 * within the related MobxStoreProvider.
 *
 * @param {string} [storeId]
 *
 *      Current state of application store
 */
function useStore(storeId = 'default') {
  const manager = getMobxManager(storeId)

  const isTestEnvironmentWithoutProviderComponent = manager == null
  if (isTestEnvironmentWithoutProviderComponent) {
    const dummyStoreForTesting = createApplicationStore()
    return dummyStoreForTesting
  }

  return manager.getContextHook()
}
