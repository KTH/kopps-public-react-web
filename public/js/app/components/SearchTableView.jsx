import React, { useState } from 'react'

import { observer } from 'mobx-react'
import { Col, Row } from 'reactstrap'

import { Link, PageHeading, SortableTable } from '@kth/kth-reactstrap/dist/components/studinfo'
import Article from '../components/Article'
import { useStore } from '../mobx'

import translate from '../../../../domain/translate'
// import { ErrorBoundary } from 'react-error-boundary'

import { courseLink } from '../util/links'
import { formatShortTerm } from '../../../../domain/term'

function codeCell(code) {
  const { language } = useStore()
  return {
    content: <Link href={courseLink(code, language)}>{code}</Link>,
    sortKey: code,
  }
}

function titleCell(code, title) {
  const { language } = useStore()
  return {
    content: <Link href={courseLink(code, language)}>{title}</Link>,
    sortKey: title,
  }
}

function compareCoursesBy(key) {
  return function compare(a, b) {
    if (a[key] < b[key]) {
      return -1
    }
    if (a[key] > b[key]) {
      return 1
    }
    return 0
  }
}

function sortAndParseByCourseCode(courses) {
  courses.sort(compareCoursesBy('courseCode'))
  const parsedCourses = courses.map(
    ({ courseCode: code, title, credits, creditUnitAbbr, educationalLevel: level, startPeriod, startTerm }) => [
      codeCell(code),
      titleCell(code, title),
      `${credits} ${creditUnitAbbr}`,
      level,
      `P${startPeriod} ${formatShortTerm(startTerm)}`,
      //P5 VT22
      // P5 spring 21
    ]
  )
  return parsedCourses
}

const SearchTableView = ({ unsortedSearchResults }) => {
  const { language } = useStore()

  const t = translate(language)

  const headers = [
    t('department_course_code'),
    t('department_course_name'),
    t('department_course_credits'),
    t('department_course_educational_level'),
    t('department_period_abbr'),
  ]
  const { searchHits } = unsortedSearchResults
  const flattCoursesArr = searchHits.map(({ course, searchHitInterval }) => ({ ...course, ...searchHitInterval }))
  const courses = sortAndParseByCourseCode(flattCoursesArr)
  const hitsNumber = courses.length
  return (
    <Row>
      <Col>
        <Article>
          {language === 'en' ? (
            <p>
              Your search returned <b>{hitsNumber}</b> result(s).
            </p>
          ) : (
            <p>
              Din s√∂kning gav <b>{hitsNumber}</b> resultat.
            </p>
          )}
          <SortableTable headers={headers} data={courses} />
        </Article>
      </Col>
    </Row>
  )
}
export default SearchTableView
