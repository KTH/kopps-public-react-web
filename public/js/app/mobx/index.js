/* eslint no-use-before-define: ["error", "nofunc"] */

import MobxStoreProvider from './MobxStoreProvider'
import useStore from './useStore'
import { compressStoreIntoJavascriptCode, uncompressStoreInPlaceFromDocument } from './compress'
// import { compressStoreIntoJavascriptCode, uncompressStoreInPlaceFromDocument } from './compressWithPako'

export { MobxStoreProvider, useStore, compressStoreIntoJavascriptCode, uncompressStoreInPlaceFromDocument }
