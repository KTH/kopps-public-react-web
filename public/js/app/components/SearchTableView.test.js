import React from 'react'
import { render, screen, within } from '@testing-library/react'
import '@testing-library/jest-dom'
import SearchTableView from './SearchTableView'
import { useStore } from '../mobx'
import {
  EXPECTED_TEST_SEARCH_HITS_MIXED_EN,
  EXPECTED_TEST_SEARCH_HITS_MIXED_SV,
  EXPECTED_TEST_SEARCH_HITS_MIXED_PERIODS_TEXTS_EN,
  EXPECTED_TEST_SEARCH_HITS_MIXED_PERIODS_TEXTS_SV,
  TEST_SEARCH_HITS_MIXED_EN,
  TEST_SEARCH_HITS_MIXED_SV,
  TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_SV,
  TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_EN,
} from './mocks/mockSearchHits'

jest.mock('../mobx')

const reseacrhHitsColHeaders = {
  en: ['Course code', 'Course name', 'Scope', 'Educational level'],
  sv: ['Kurskod', 'Kursnamn', 'Omfattning', 'Utbildningsnivå'],
}
const mixedHitsColHeaders = {
  en: [...reseacrhHitsColHeaders.en, 'Periods'],
  sv: [...reseacrhHitsColHeaders.sv, 'Perioder'],
}
const eduLevelTranslations = {
  EN: { PREPARATORY: 'Pre-university level', BASIC: 'First cycle', ADVANCED: 'Second cycle', RESEARCH: 'Third cycle' },
  SV: { PREPARATORY: 'Förberedande nivå', BASIC: 'Grundnivå', ADVANCED: 'Avancerad nivå', RESEARCH: 'Forskarnivå' },
}

describe('Component <SearchTableView> for RESEARCH courses', () => {
  test('creates a table with 4 columns for RESEARCH courses (without column for period intervals). English. 1A', () => {
    useStore.mockReturnValue({ language: 'en', languageIndex: 0 })
    const { asFragment } = render(
      <SearchTableView unsortedSearchResults={TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_EN} />
    )

    expect(asFragment()).toMatchSnapshot()
    // or more detailed
    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Your search returned 3 result(s).')

    const rows = screen.queryAllByRole('row')
    const columnHeaders = screen.getAllByRole('columnheader')
    const contentCells = screen.queryAllByRole('cell')
    expect(rows).toHaveLength(4)
    expect(columnHeaders).toHaveLength(4)
    expect(contentCells).toHaveLength(12)

    try {
      columnHeaders.map((colHeader, index) => expect(colHeader).toHaveTextContent(reseacrhHitsColHeaders.en[index]))
    } catch (error) {
      error.message = `${`Table head row missing a correct translations. List of correct translations ${reseacrhHitsColHeaders.en.join(
        ', '
      )}`}\n\n ${error}`

      throw error
    }

    try {
      rows.slice(1).forEach((row, index) => {
        const utils = within(row)
        const { course } = TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_EN.searchHits[index]
        expect(utils.getAllByRole('cell')[0]).toHaveTextContent(course.courseCode, { exact: true })
        expect(utils.getAllByRole('cell')[1]).toHaveTextContent(course.title, { exact: true })
        expect(utils.getAllByRole('cell')[2]).toHaveTextContent(`${course.credits} ${course.creditUnitAbbr}`, {
          exact: true,
        })
        expect(utils.getAllByRole('cell')[3]).toHaveTextContent('Third cycle')
      })
    } catch (error) {
      error.message = `${`Courses has been rendered incorrect. Course content in the table is differ from an expected content`}\n\n ${error}`

      throw error
    }
  })

  test('creates a table with 4 columns for RESEARCH courses (without column for period intervals). Swedish. 2A', () => {
    useStore.mockReturnValue({ language: 'sv', languageIndex: 1 })
    const { asFragment } = render(
      <SearchTableView unsortedSearchResults={TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_SV} />
    )

    expect(asFragment()).toMatchSnapshot()
    // // or
    // or more detailed
    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Din sökning gav 3 resultat.')

    const rows = screen.queryAllByRole('row')
    const columnHeaders = screen.getAllByRole('columnheader')
    const contentCells = screen.queryAllByRole('cell')
    expect(rows).toHaveLength(4)
    expect(columnHeaders).toHaveLength(4)
    expect(contentCells).toHaveLength(12)

    try {
      columnHeaders.map((colHeader, index) => expect(colHeader).toHaveTextContent(reseacrhHitsColHeaders.sv[index]))
    } catch (error) {
      error.message = `${`Table head row missing a correct translations. List of correct translations ${reseacrhHitsColHeaders.sv.join(
        ', '
      )}`}\n\n ${error}`

      throw error
    }

    try {
      rows.slice(1).forEach((row, index) => {
        const utils = within(row)
        const { course } = TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_SV.searchHits[index]
        expect(utils.getAllByRole('cell')[0]).toHaveTextContent(course.courseCode, { exact: true })
        expect(utils.getAllByRole('cell')[1]).toHaveTextContent(course.title, { exact: true })
        expect(utils.getAllByRole('cell')[2]).toHaveTextContent(`${course.credits} ${course.creditUnitAbbr}`, {
          exact: true,
        })
        expect(utils.getAllByRole('cell')[3]).toHaveTextContent('Forskarnivå')
      })
    } catch (error) {
      error.message = `${`Courses has been rendered incorrect. Course content in the table is differ from an expected content`}\n\n ${error}`

      throw error
    }
  })

  test('creates a table with 5 columns for mixed types of courses (include column for period intervals). English. 1B', () => {
    useStore.mockReturnValue({ language: 'en', languageIndex: 0 })
    const { asFragment } = render(
      <SearchTableView unsortedSearchResults={TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_EN} />
    )

    expect(asFragment()).toMatchSnapshot()
    // or more detailed
    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Your search returned 3 result(s).')

    const rows = screen.queryAllByRole('row')
    const columnHeaders = screen.getAllByRole('columnheader')
    const contentCells = screen.queryAllByRole('cell')
    expect(rows).toHaveLength(4)
    expect(columnHeaders).toHaveLength(4)
    expect(contentCells).toHaveLength(12)

    try {
      columnHeaders.map((colHeader, index) => expect(colHeader).toHaveTextContent(reseacrhHitsColHeaders.en[index]))
    } catch (error) {
      error.message = `${`Table head row missing a correct translations. List of correct translations ${reseacrhHitsColHeaders.en.join(
        ', '
      )}`}\n\n ${error}`

      throw error
    }

    try {
      rows.slice(1).forEach((row, index) => {
        const utils = within(row)
        const { course } = TEST_SEARCH_RESEARCH_THIRD_CYCLE_COURSES_EN.searchHits[index]
        expect(utils.getAllByRole('cell')[0]).toHaveTextContent(course.courseCode, { exact: true })
        expect(utils.getAllByRole('cell')[1]).toHaveTextContent(course.title, { exact: true })
        expect(utils.getAllByRole('cell')[2]).toHaveTextContent(`${course.credits} ${course.creditUnitAbbr}`, {
          exact: true,
        })
        expect(utils.getAllByRole('cell')[3]).toHaveTextContent('Third cycle')
      })
    } catch (error) {
      error.message = `${`Courses has been rendered incorrect. Course content in the table is differ from an expected content`}\n\n ${error}`

      throw error
    }
  })
})

describe('Component <SearchTableView> for MIXED types of courses', () => {
  test('creates a table with 5 columns for mixed types of courses (include column for period intervals). English. 1B', () => {
    useStore.mockReturnValue({ language: 'en', languageIndex: 0 })
    const { asFragment } = render(<SearchTableView unsortedSearchResults={TEST_SEARCH_HITS_MIXED_EN} />)

    expect(asFragment()).toMatchSnapshot()
    // or more detailed
    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Your search returned 5 result(s).')

    const rows = screen.queryAllByRole('row')
    const columnHeaders = screen.getAllByRole('columnheader')
    const contentCells = screen.queryAllByRole('cell')
    expect(rows).toHaveLength(6)
    expect(columnHeaders).toHaveLength(5)
    expect(contentCells).toHaveLength(25)

    try {
      columnHeaders.map((colHeader, index) => expect(colHeader).toHaveTextContent(mixedHitsColHeaders.en[index]))
    } catch (error) {
      error.message = `${`Table head row missing a correct translations. List of correct translations ${mixedHitsColHeaders.en.join(
        ', '
      )}`}\n\n ${error}`

      throw error
    }

    try {
      rows.slice(1).forEach((row, index) => {
        const utils = within(row)
        const { course } = EXPECTED_TEST_SEARCH_HITS_MIXED_EN.searchHits[index]
        expect(utils.getAllByRole('cell')[0]).toHaveTextContent(course.courseCode, { exact: true })
        expect(utils.getAllByRole('cell')[1]).toHaveTextContent(course.title, { exact: true })
        expect(utils.getAllByRole('cell')[2]).toHaveTextContent(`${course.credits} ${course.creditUnitAbbr}`, {
          exact: true,
        })
        expect(utils.getAllByRole('cell')[3]).toHaveTextContent(
          course.educationalLevel ? eduLevelTranslations.EN[course.educationalLevel] : '',
          { exact: true }
        )
        expect(utils.getAllByRole('cell')[4]).toHaveTextContent(
          EXPECTED_TEST_SEARCH_HITS_MIXED_PERIODS_TEXTS_EN[index],
          { exact: true }
        )
      })
    } catch (error) {
      error.message = `${`Courses has been rendered incorrect. Course content in the table is differ from an expected content`}\n\n ${error}`

      throw error
    }
  })

  test('creates a table with 5 columns for MIXED type of courses (include a column for period intervals). Swedish. 2B', () => {
    useStore.mockReturnValue({ language: 'sv', languageIndex: 1 })
    const { asFragment } = render(<SearchTableView unsortedSearchResults={TEST_SEARCH_HITS_MIXED_SV} />)

    expect(asFragment()).toMatchSnapshot()
    // // or
    // or more detailed
    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Din sökning gav 5 resultat.')

    const rows = screen.queryAllByRole('row')
    const columnHeaders = screen.getAllByRole('columnheader')
    const contentCells = screen.queryAllByRole('cell')
    expect(rows).toHaveLength(6)
    expect(columnHeaders).toHaveLength(5)
    expect(contentCells).toHaveLength(25)

    try {
      columnHeaders.map((colHeader, index) => expect(colHeader).toHaveTextContent(mixedHitsColHeaders.sv[index]))
    } catch (error) {
      error.message = `${`Table head row missing a correct translations. List of correct translations ${mixedHitsColHeaders.sv.join(
        ', '
      )}`}\n\n ${error}`

      throw error
    }

    try {
      rows.slice(1).forEach((row, index) => {
        const utils = within(row)
        const { course } = EXPECTED_TEST_SEARCH_HITS_MIXED_SV.searchHits[index]
        expect(utils.getAllByRole('cell')[0]).toHaveTextContent(course.courseCode, { exact: true })
        expect(utils.getAllByRole('cell')[1]).toHaveTextContent(course.title, { exact: true })
        expect(utils.getAllByRole('cell')[2]).toHaveTextContent(`${course.credits} ${course.creditUnitAbbr}`, {
          exact: true,
        })
        expect(utils.getAllByRole('cell')[3]).toHaveTextContent(
          course.educationalLevel ? eduLevelTranslations.SV[course.educationalLevel] : '',
          { exact: true }
        )
        expect(utils.getAllByRole('cell')[4]).toHaveTextContent(
          EXPECTED_TEST_SEARCH_HITS_MIXED_PERIODS_TEXTS_SV[index],
          { exact: true }
        )
      })
    } catch (error) {
      error.message = `${`Courses has been rendered incorrectly. Course content in the table is differ from an expected content`}\n\n ${error}`

      throw error
    }
  })
})
