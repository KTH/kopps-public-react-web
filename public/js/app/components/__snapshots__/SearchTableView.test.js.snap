// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <SearchTableView> for MIXED types of courses creates a table with 5 columns for MIXED type of courses (include a column for period intervals). Swedish. 2B 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Din sökning gav 
            <b>
              5
            </b>
             resultat.
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Kurskod
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Kursnamn
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Omfattning
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Utbildningsnivå
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Perioder
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AF233X?startterm=20212"
                  >
                    AF233X
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AF233X?startterm=20212"
                  >
                    Examensarbete inom byggnadsmateriallära, avancerad nivå
                  </a>
                </td>
                <td>
                  30 hp
                </td>
                <td>
                  Avancerad nivå
                </td>
                <td>
                  P1 HT21 - P2 HT21
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AF2402?startterm=20212"
                  >
                    AF2402
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AF2402?startterm=20212"
                  >
                    Akustik och brand
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Grundnivå
                </td>
                <td>
                  P1 HT21 - P3 VT22
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AH2905?startterm=20212"
                  >
                    AH2905
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AH2905?startterm=20212"
                  >
                    Avancerad analys och design av vägbeläggningar
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td />
                <td>
                  P1 HT21
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3901"
                  >
                    FAF3901
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3901"
                  >
                    Avancerad reologi för bituminösa material
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Forskarnivå
                </td>
                <td />
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAH3904?startterm=20212"
                  >
                    FAH3904
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAH3904?startterm=20212"
                  >
                    Introduktion till asfaltskemin
                  </a>
                </td>
                <td>
                  4 hp
                </td>
                <td>
                  Förberedande nivå
                </td>
                <td>
                  P0 HT21 - P5 VT22
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <SearchTableView> for MIXED types of courses creates a table with 5 columns for mixed types of courses (include column for period intervals). English. 1B 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Your search returned 
            <b>
              5
            </b>
             result(s).
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course code
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course name
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Scope
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Educational level
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Periods
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AF233X?startterm=20212&l=en"
                  >
                    AF233X
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AF233X?startterm=20212&l=en"
                  >
                    Degree Project in Building Materials, Second Cycle
                  </a>
                </td>
                <td>
                  30 hp
                </td>
                <td>
                  Second cycle
                </td>
                <td>
                  P1 Autumn 21 - P2 Autumn 21
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AF2402?startterm=20212&l=en"
                  >
                    AF2402
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AF2402?startterm=20212&l=en"
                  >
                    Acoustics and Fire
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  First cycle
                </td>
                <td>
                  P1 Autumn 21 - P3 Spring 22
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AH2905?startterm=20212&l=en"
                  >
                    AH2905
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AH2905?startterm=20212&l=en"
                  >
                    Advanced Pavement Engineering Analysis and Design
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td />
                <td>
                  P1 Autumn 21
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3901?l=en"
                  >
                    FAF3901
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3901?l=en"
                  >
                    Advanced Rheology of Bituminous Materials
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Third cycle
                </td>
                <td />
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAH3904?startterm=20212&l=en"
                  >
                    FAH3904
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAH3904?startterm=20212&l=en"
                  >
                    Introduction to Asphalt Chemistry
                  </a>
                </td>
                <td>
                  4 hp
                </td>
                <td>
                  Pre-university level
                </td>
                <td>
                  P0 Autumn 21 - P5 Spring 22
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <SearchTableView> for RESEARCH courses creates a table with 4 columns for RESEARCH courses (without column for period intervals). English. 1A 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Your search returned 
            <b>
              3
            </b>
             result(s).
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course code
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course name
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Scope
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Educational level
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3302?l=en"
                  >
                    FAF3302
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3302?l=en"
                  >
                    Project in Building Materials Technology
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Third cycle
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3304?l=en"
                  >
                    FAF3304
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3304?l=en"
                  >
                    Wood Chemistry, Biocomposites and Building Materials
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Third cycle
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3305?l=en"
                  >
                    FAF3305
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3305?l=en"
                  >
                    Pavement Design and Performance Prediction
                  </a>
                </td>
                <td>
                  3 hp
                </td>
                <td>
                  Third cycle
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <SearchTableView> for RESEARCH courses creates a table with 4 columns for RESEARCH courses (without column for period intervals). Swedish. 2A 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Din sökning gav 
            <b>
              3
            </b>
             resultat.
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Kurskod
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Kursnamn
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Omfattning
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Utbildningsnivå
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3302"
                  >
                    FAF3302
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3302"
                  >
                    Projekt i byggnadsmaterialteknik
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Forskarnivå
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3304"
                  >
                    FAF3304
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3304"
                  >
                    Träkemi för biokompositer som byggnadsmaterial
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Forskarnivå
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3305"
                  >
                    FAF3305
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3305"
                  >
                    Vägdimensionering och prestandautvärdering
                  </a>
                </td>
                <td>
                  3 hp
                </td>
                <td>
                  Forskarnivå
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <SearchTableView> for RESEARCH courses creates a table with 5 columns for mixed types of courses (include column for period intervals). English. 1B 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Your search returned 
            <b>
              3
            </b>
             result(s).
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course code
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course name
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Scope
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Educational level
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3302?l=en"
                  >
                    FAF3302
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3302?l=en"
                  >
                    Project in Building Materials Technology
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Third cycle
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3304?l=en"
                  >
                    FAF3304
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3304?l=en"
                  >
                    Wood Chemistry, Biocomposites and Building Materials
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  Third cycle
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3305?l=en"
                  >
                    FAF3305
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/FAF3305?l=en"
                  >
                    Pavement Design and Performance Prediction
                  </a>
                </td>
                <td>
                  3 hp
                </td>
                <td>
                  Third cycle
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
